Feat proof versioning

thomas-waite opened the pull request:
## Summary
Implement a proof construction versioning system in the `aztec.js` package. 

## Description
The motivation of this PR was to develop a systematic method whereby additional proofs could be added to `aztec.js`, without making any breaking changes to the API. 

### Individual proof export scripts
Each proof now has it's own proof export script. The main method in each script is a function with the same name as the proof, to prevent any breaking changes. The API to construct a proof is the same as before:
```
const { JoinSplitProof } = require('aztec.js');

const proof = new JoinSplitProof(...args);
```

There is however now also the option to specify a version of a proof from a particular epoch, such as:

```
const proof = new JoinSplitProof.epoch(2)(...args);
```

### proofHandler.js
This is a generic module that is used to manage the process of exporting individual proofs, and is relied on for the 'backend' of each individual proof export script. 

It also enables additional functionality that allows a user to specify a `defaultEpochNum` which a user wants to use proofs from. 

### proofCatalog.js
Closely related to the `proofHandler` and required for it's functionality is the `proofCatalog.js` script. This contains a mapping between epoch numbers and the proof versions, which is queried by the `proofHandler` when the user calls a proof.

Using such a mapping to keep track of which proofs belong to which epoch numbers will make this versioning system more scalable in the future - additional proofs and epochs can easily be added.  

<!--- Describe your changes in detail -->

## Testing instructions
It is important that the `proofHandler` behaves as expected. As such various success and failure tests have been added to confirm behaviour.

thomas-waite committed:
feat(aztec.js): implement proof versioning dir structure

aztec.js proof construction versioning to follow same uint24 _proof structure as protocol package

thomas-waite committed:
feat(dev-utils): add proof and error constants

thomas-waite committed:
feat(aztec.js): add proof construction versioning export scripts
add following functionality:
set a default epoch number for a proof to be exported from
add ability to use a proof from a particular epoch, selected by the user

thomas-waite committed:
refactor(aztec.js): add proofExport.js to reduce WET code

thomas-waite committed:
feat(aztec.js): add proofHandler, to manage proof exporting

thomas-waite committed:
feat(aztec.js): add proofCatalog, to manage proof versions

thomas-waite committed:
refactor(aztec.js): update proof export scripts with proofHandler

thomas-waite committed:
test(aztec.js): test proofHandler behaviour

thomas-waite committed:
refactor(dev-utils): remove now redundant constants
constants now replaced with proofCatalog in aztec.js

thomas-waite committed:
docs(aztec.js): update function docs

ArnSch commented on packages/aztec.js/src/proof/proofCatalog.js:
```diff
@@ -0,0 +1,48 @@
+/*
+ * Default epoch number for which proof construction methods are exported
+ */
+proofCatalog.defaultEpochNum = 1;
```
The way this is currently structured, it assumes that we will never add an epoch on one proof without also adding proofs for all the existing proof types for that new epoch. Is that true?

thomas-waite replied:
As discussed earlier, good point. We're going to add another mapping whereby each proof type has a default proof epoch, which the user can set

ArnSch commented on packages/aztec.js/test/proof/dividend/verifier.js:
```diff
@@ -7,7 +7,7 @@ const { padLeft, randomHex } = require('web3-utils');
 
 const { DividendProof } = require('../../../src/proof');
-const DividendVerifier = require('../../../src/proof/dividend/verifier');
+const DividendVerifier = require('../../../src/proof/epoch0/UTILITY/dividend/verifier');
```
small, but I think it makes sense for the proof type to be a parent of an epoch:

```
- UTILITY
  - epoch0
  - epoch1
```
instead of 
```
- Epoch0
  - utility
  - balanced
- Epoch1
  - utility
```

thomas-waite replied:
As discussed, this makes sense

ArnSch commented on packages/aztec.js/test/proof/joinSplitFluid/verifier.js:
```diff
@@ -1,112 +1,112 @@
-        });
-    });
-});
+// const { errors } = require('@aztec/dev-utils');
```
Do you mean for this to be commented?

thomas-waite replied:
Good point, not sure how I missed that - should be uncommented. Fixed

ArnSch reviewed (changes requested):
Small comments around relative hierarchy of proof types vs epochs, but otherwise 👍

thomas-waite committed:
test(aztec.js): uncomment test, fix imports

thomas-waite committed:
refactor(aztec.js): implement PR feedback - change proof dir structure

proof type is now a child of epoch number in the proof module
proof base classes have been broken out into a seperate folder, organised by the proof epoch number they correspond to

thomas-waite committed:
feat(aztec.js): add defaultProofEpochNums per proof

thomas-waite committed:
test(aztec.js): confirm defaultProofEpochNums behaviour

thomas-waite commented:
Ready for a re-check

Commit:
fix(aztec.js): rejig proof versioning to clean up api

Commit:
docs(aztec.js): rename boolean
